lib/segment/src/types.rs

```diff
@@ -345,6 +345,23 @@ pub struct HnswConfig {
     pub payload_m: Option<usize>,
 }
 
+impl HnswConfig {
+    /// Detect configuration mismatch against `other` that requires rebuilding
+    ///
+    /// Returns true only if both conditions are met:
+    /// - this configuration does not match `other`
+    /// - to effectively change the configuration, a HNSW rebuild is required
+    ///
+    /// For example, a change in `max_indexing_threads` will not require rebuilding because it
+    /// doesn't affect the final index, and thus this would return false.
+    pub fn mismatch_requires_rebuild(&self, other: &Self) -> bool {
+        self.m != other.m
+            || self.ef_construct != other.ef_construct
+            || self.on_disk != other.on_disk
```

Comment:
If only `on_disk` was toggled, do we actually need to do full rebuild?
Can't we reuse the existing index data and change the way it is loaded?

Reply:
I don't think the two variants are compatible. But this is good to investigate, thanks!

Reply:
It is compatible. We even have a unit test for it: https://github.com/qdrant/qdrant/blob/b13df2272cb351986d2a47844127a5280dfbc3cb/lib/segment/src/index/hnsw_index/graph_links.rs#L703

Thanks @IvanPleshkov for pointing me to it. :raised_hands:

Reply:
Fixed in <https://github.com/qdrant/qdrant/pull/2083/commits/c1a72f0e1b04a9a4b724a2b12c0cc7d653530688>.

Reply:
During a call with Andrey we agreed to go back to the the naive approach of reoptimizing the whole collection again. This isn't intended for production use so performance isn't too much of a concern here. We can always optimize this in the future if we so desire.

So I'll revert this change.